h2integrate/postprocess/mapping.py

```diff
@@ -0,0 +1,852 @@
+
+    # Add basemap tiles with contextily
+    ctx.add_basemap(
+        ax,
+        crs=map_preferences.web_map_crs,
+        source=map_preferences.basemap_provider,
+        zoom=map_preferences.basemap_zoom,
+    )
+
+    # NOTE: when plotting multiple layers, set this to True only when plotting the last layer
+    if show_plot:
+        plt.show()
+
+    # Save figure if save_plot_fpath is present
+    if save_plot_fpath is not None:
+        if isinstance(save_plot_fpath, str):
+            save_plot_fpath = Path(save_plot_fpath)
+        fig.savefig(save_plot_fpath)
+
+    return fig, ax, shipping_route_gdf
+
+
+def calculate_geodataframe_total_bounds(*gdfs: gpd.GeoDataFrame | list[gpd.GeoDataFrame]):
+    """
+    Calculate the combined spatial bounds for one or more GeoDataFrames.
+
+    Computes the minimum and maximum X and Y coordinates across all provided GeoDataFrames and
+    returns both the absolute bounds and the corresponding coordinate ranges.
+    All GeoDataFrames must share the same coordinate reference system (CRS).
+
+    Args:
+        *gdfs (gpd.GeoDataFrame | list[gpd.GeoDataFrame]): One or more GeoDataFrame objects for
+            which to compute the combined spatial extent.
+
+    Returns:
+        coord_range_dict (dict): A dictionary containing the combined bounds and coordinate ranges
+            with the following keys:
+
+            - ``min_x``: Minimum X coordinate across all GeoDataFrames
+            - ``min_y``: Minimum Y coordinate across all GeoDataFrames
+            - ``max_x``: Maximum X coordinate across all GeoDataFrames
+            - ``max_y``: Maximum Y coordinate across all GeoDataFrames
+            - ``x_range``: Total range in the X direction (max_x - min_x)
+            - ``y_range``: Total range in the Y direction (max_y - min_y)
+
+    Raises:
+        ValueError: If no GeoDataFrames are provided or if the GeoDataFrames do not all share the
+            same CRS.
+
+    """
+
+    # raise error if no gdf is provided as keyword argument
+    if not gdfs:
+        raise ValueError("Must provide at least one GeoDataFrame.")
+
+    # Validate all gdfs are in the same CRS
+    base_crs = gdfs[0].crs
+    for gdf in gdfs:
+        if gdf.crs != base_crs:
+            raise ValueError("All GeoDataFrames must have the same CRS.")
+
+    # Extract the min and max X (Longitude) and Y (Latitude) from the total bounds of all gdfs
+    min_xs, min_ys, max_xs, max_ys = zip(*(gdf.total_bounds for gdf in gdfs))
+
+    min_x = min(min_xs)
+    min_y = min(min_ys)
+    max_x = max(max_xs)
+    max_y = max(max_ys)
+    x_range = max_x - min_x
+    y_range = max_y - min_y
+
+    # Construct dictionary with relevant values
+    coord_range_dict = {
+        "min_x": min_x,
+        "min_y": min_y,
+        "max_x": max_x,
+        "max_y": max_y,
+        "x_range": x_range,
+        "y_range": y_range,
+    }
+
+    return coord_range_dict
+
+
+def auto_detect_lat_long_columns(results_df: pd.DataFrame):
```

Comment:
I would recommend adding a keyword argument called `which` that accepts one of "lat", "lon", or "both", or some other variation of those three options because this is called multiple times in a row to handle None inputs for one column, and then the next. Rather than taking a more involved control flow where this is used, it seems easier to allow for capturing one or the other, or both column names. The control flow should then be similar to the following.

```python
latitude_var_name = None
longitude_var_name = None
if which not in ("lat", "lon", "both"):
    raise ValueError(...)

if which in ("lat", "both"):
    ... # latitude logic

if which in ("lon", "both"):
    ... # longitude logic

return latitude_var_name, longitude_var_name
```